src/Data/KleeneList/Base.agda

```diff
@@ -0,0 +1,366 @@
+  mapMaybe⋆ (∹ xs) = mapMaybe⁺ xs
+
+------------------------------------------------------------------------
+-- Applicative Operations
+
+pure⁺ : A → A ⁺
+head (pure⁺ x) = x
+tail (pure⁺ x) = []
+
+pure⋆ : A → A ⋆
+pure⋆ x = ∹ pure⁺ x
+
+module Apply where
+  _⋆<*>⋆_ : (A → B) ⋆ → A ⋆ → B ⋆
+  _⁺<*>⋆_ : (A → B) ⁺ → A ⋆ → B ⋆
+  _⋆<*>⁺_ : (A → B) ⋆ → A ⁺ → B ⋆
+  _⁺<*>⁺_ : (A → B) ⁺ → A ⁺ → B ⁺
+
+  []     ⋆<*>⋆ xs = []
+  (∹ fs) ⋆<*>⋆ xs = fs ⁺<*>⋆ xs
+
+  fs ⁺<*>⋆ xs = map⋆ (head fs) xs ++⋆ (tail fs ⋆<*>⋆ xs)
+
+  []     ⋆<*>⁺ xs = []
+  (∹ fs) ⋆<*>⁺ xs = ∹ fs ⁺<*>⁺ xs
+
+  fs ⁺<*>⁺ xs = map⁺ (head fs) xs Concat.⁺++⋆ (tail fs ⋆<*>⁺ xs)
+open Apply public using () renaming (_⋆<*>⋆_ to _<*>⋆_; _⁺<*>⁺_ to _<*>⁺_)
+
+------------------------------------------------------------------------
+-- Monadic Operations
+
+module Bind where
+  _⁺>>=⁺_ : A ⁺ → (A → B ⁺) → B ⁺
+  _⁺>>=⋆_ : A ⁺ → (A → B ⋆) → B ⋆
+  _⋆>>=⁺_ : A ⋆ → (A → B ⁺) → B ⋆
+  _⋆>>=⋆_ : A ⋆ → (A → B ⋆) → B ⋆
+
+  (x & xs) ⁺>>=⁺ k = k x Concat.⁺++⋆ (xs ⋆>>=⁺ k)
+
+  (x & xs) ⁺>>=⋆ k = k x Concat.⋆++⋆ (xs ⋆>>=⋆ k)
+
+  []     ⋆>>=⋆ k = []
+  (∹ xs) ⋆>>=⋆ k = xs ⁺>>=⋆ k
+
+  []     ⋆>>=⁺ k = []
+  (∹ xs) ⋆>>=⁺ k = ∹ xs ⁺>>=⁺ k
+open Bind public using  () renaming (_⋆>>=⋆_ to _>>=⋆_; _⁺>>=⁺_ to _>>=⁺_)
+
+------------------------------------------------------------------------
+-- Scans
+
+module Scanr (f : A → B → B) (b : B) where
+  cons : A → B ⁺ → B ⁺
+  head (cons x xs) = f x (head xs)
+  tail (cons x xs) = ∹ xs
+
+  scanr⁺ : A ⁺ → B ⁺
+  scanr⋆ : A ⋆ → B ⁺
+
+  scanr⋆ = foldr⋆ cons (b & [])
+  scanr⁺ = foldr⁺ cons (b & [])
+open Scanr public using (scanr⁺; scanr⋆)
+
+module _ (f : B → A → B) where
+  scanl⋆ : B → A ⋆ → B ⁺
+  head (scanl⋆ b xs) = b
+  tail (scanl⋆ b []) = []
+  tail (scanl⋆ b (∹ xs)) = ∹ scanl⋆ (f b (head xs)) (tail xs)
+
+  scanl⁺ : B → A ⁺ → B ⁺
+  head (scanl⁺ b xs) = b
+  tail (scanl⁺ b xs) = ∹ scanl⋆ (f b (head xs)) (tail xs)
+
+
+  scanl₁ : B → A ⁺ → B ⁺
+  scanl₁ b xs = scanl⋆ (f b (head xs)) (tail xs)
+
+------------------------------------------------------------------------
+-- Accumulating maps
+
+module _ (f : B → A → (B × C)) where
+  mapAccumˡ⋆ : B → A ⋆ → (B × C ⋆)
+  mapAccumˡ⁺ : B → A ⁺ → (B × C ⁺)
+
+  mapAccumˡ⋆ b [] = b , []
+  mapAccumˡ⋆ b (∹ xs) = map₂ ∹_ (mapAccumˡ⁺ b xs)
+
+  mapAccumˡ⁺ b (x & xs) =
+    let y , ys = f b x
+        z , zs = mapAccumˡ⋆ y xs
+    in z , ys & zs
+
+module _ (f : A → B → (C × B)) (b : B) where
+  mapAccumʳ⋆ : A ⋆ → (C ⋆ × B)
+  mapAccumʳ⁺ : A ⁺ → (C ⁺ × B)
+
+  mapAccumʳ⋆ [] = [] , b
+  mapAccumʳ⋆ (∹ xs) = map₁ ∹_ (mapAccumʳ⁺ xs)
+
+  mapAccumʳ⁺ (x & xs) =
+    let ys , y = mapAccumʳ⋆ xs
+        zs , z = f x y
+    in zs & ys , z
+
+------------------------------------------------------------------------
+-- Non-Empty Folds
+
+last : A ⁺ → A
+last (x & []) = x
+last (_ & ∹ xs) = last xs
+
+module _ (f : A → A → A) where
+  foldr1 : A ⁺ → A
+  foldr1 (x & []) = x
+  foldr1 (x & ∹ xs) = f x (foldr1 xs)
```

Comment:
Why is there a 1 at the end of this name?

Reply:
It's taken from the [Haskell function of the same name](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#v:foldr1)

Reply:
We also use this name in [`Data.List.NonEmpty`](https://agda.github.io/agda-stdlib/Data.List.NonEmpty.html#2849) (although the 1 there
is an index, which should also be the case here for consistency)

Reply:
Oh, I see it's already been fixed!